Add support for Reciprocal Rank Fusion to the search API

jdconrad opened the pull request:
This change at a high level adds global ranking on the coordinating node at the end of query reduction prior to the fetch phase. Individual rank methods are defined in plugins.

The first rank plugin added as part of this change is reciprocal rank fusion (RRF). RRF uses a relatively simple formula for merging `1...n` results sets together with `sum(1/(k+d))` where k is a ranking constant and d is a document's scored position within a result set from a query.

The API for this change adds a `rank` top-level element to the search endpoint. And for RRF an `rrf element` under the `rank` element. An example:

```
{
  "query": {
    "match": {
      "product": {
        "query": "brown shoes"
      }
    }
  },
  "knn": {
    "field": "product-vector",
    "query_vector": [54, 10, -2],
    "k": 20,
    "num_candidates": 75
  },
  "rank": {
     "rrf": {
        "window_size": 100,
        "rank_constant": 20
     }
  }
}
```

The above example will execute two separate queries for the match query and the kNN search. It will preserve separate result sets up the point where the queries are ranked on the coordinating node using RRF.

Ranking only supports a single query and 1-to-many knn searches or no query and 2-to-many knn searches. 

Ranking currently does not support the following:
* sort
* scroll
* rescore
* collapse fields
* suggestions
* highlighting
* point in time
* profiling
* explain

Follow up:
* Documentation will be completed as a separate follow-up PR.
* Add ability to have multiple standard queries as part of ranking.

jdconrad committed:
Split out single search data from global search data in
QuerySearchResult.

jdconrad committed:
fix from setter

jdconrad committed:
add RerankQueryBuilder

jdconrad committed:
add internal rrf reranker for secondary results

jdconrad committed:
add rerankbuilder

jdconrad committed:
add rerank initial tests

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
revert KnnSearchSingleNodeTests

jdconrad committed:
add reranker to SearchContext

jdconrad committed:
fix tests

jdconrad committed:
update to plumb through size and window_size; also do not use sort on
knn queries

jdconrad committed:
add basic yaml test for rrf

jdconrad committed:
style fixes

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix writeTo

jdconrad committed:
register named query

jdconrad committed:
more fixes

jdconrad committed:
fix serialization bug

jdconrad committed:
fix another serialization bug

benwtrent commented on server/src/main/java/org/elasticsearch/search/query/QuerySearchResult.java:
```diff
@@ -29,18 +29,143 @@
 import static org.elasticsearch.common.lucene.Lucene.readTopDocs;
 import static org.elasticsearch.common.lucene.Lucene.writeTopDocs;
 
 public final class QuerySearchResult extends SearchPhaseResult {
```
I wonder if the "secondary results" really need to be `SingleSearchResult`. It seems like they should be something completely different that only keeps track of what the reranker needs.

To me that would remove the need of this weird subclass. 

As for how to handle the `QuerySearchResult` being singular, I think we can still do this by setting the values given the reranker steps on the QueryPhase, many of these are still "global" concepts (from, size, terminatedEarly, serviceTimeWMA, etc.)

jdconrad replied:
Completely changed this. I added a `RankShardResult` base class with an `RRFRankShardResult` that contains a list of `TopDocs` and nothing else.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rescore/RescorePhase.java:
```diff
@@ -21,8 +22,8 @@
-    public static void execute(SearchContext context) {
-        TopDocs topDocs = context.queryResult().topDocs().topDocs;
+    public static void execute(SearchContext context, SingleSearchResult singleSearchResult) {
+        TopDocs topDocs = singleSearchResult.topDocs().topDocs;
```
This should be modified IMO to accept a context and return top docs (or null) and the caller can figure out what to do with it. This may make the reranker execution stuff easier.

jdconrad replied:
This is changed as described above where it's basically just an additional list of top docs.

benwtrent commented on server/src/main/java/org/elasticsearch/search/SearchModule.java:
```diff
@@ -1119,6 +1120,14 @@ private void registerQueryParsers(List<SearchPlugin> plugins) {
             )
         );
 
+        registerQuery(
+            new QuerySpec<>(
+                RerankQueryBuilder.NAME,
+                RerankQueryBuilder::new,
+                parser -> { throw new IllegalArgumentException("[rerank] queries cannot be provided directly"); }
+            )
+        );
+
```
While this is a clever idea, I don't think this is the abstraction we want. It may be that on the individual node itself, the reranker provides something like this. But to me we should be passing around a `RerankerContext` object or something that tracks stuff instead of hijacking query.

jdconrad replied:
Changed this to pass the additional queries along as part of the `RankBuilder` using the internal transport protocol.

benwtrent reviewed (commented):
This is indeed a substantial and difficult change. I let some initial ideas. I think it would be good to catching synchronously to "rubber-duck" some ideas.

I wonder if many of the issues encountered could be fixed by:

- RerankerContext object for tracking the required things for final ranking
- A new RerankerSearchContext that overrides methods to be used in the QueryPhase
- Updating query search results to return RerankerContext
- have the search phase controller apply the "reranker" give the context and returns the ranking positions for each hit (instead of adjusting score)

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
revert QuerySearchResult

jdconrad committed:
revert QueryPhase

jdconrad committed:
revert RescorePhase

jdconrad committed:
more reversions

jdconrad committed:
revert QueryPhaseResultConsumer

jdconrad committed:
revert tests

jdconrad committed:
remove rerank query

jdconrad committed:
revert DfsQueryPhase

jdconrad committed:
clean up builders a bit

jdconrad committed:
Start creating appropriate collector

jdconrad committed:
add rrf collector

jdconrad committed:
add outline for result

jdconrad committed:
collector test

jdconrad committed:
attempt at query wrapper

jdconrad committed:
add custom query and add rank results

jdconrad committed:
start to update output to handle rank

jdconrad committed:
update SearchHit to accept rank information

jdconrad committed:
spotless

jdconrad committed:
fix test name

jdconrad committed:
Merge branch 'main' into rrf14

jdconrad committed:
add specialized rrf rank query builder

jdconrad committed:
revert QueryPhase

jdconrad committed:
revert topdocscollectorcontext

jdconrad committed:
revert QueryCollectorContext

jdconrad committed:
add rank query context to process multiple queries

jdconrad committed:
pass through size for ranking

jdconrad committed:
fix from parameter

jdconrad committed:
rename RRFRankQueryBuilder to RankQueryBuilder

jdconrad committed:
update builders to add additional information

jdconrad committed:
simple renames

jdconrad committed:
clean up builders from coordinator to shard

jdconrad committed:
spotless

jdconrad committed:
Merge branch 'main' into rrf14

jdconrad committed:
fix binary read/write issues

jdconrad commented on server/src/main/java/org/elasticsearch/search/rerank/RRFReranker.java:
```diff
@@ -0,0 +1,78 @@
+
+                    return value;
+                });
+            }
```
I have changed this to have an `_rank` in `SearchHit`.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rerank/RRFReranker.java:
```diff
@@ -0,0 +1,78 @@
+            null,
+            0
+        );
+    }
```
This now retains the sort information.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rerank/Reranker.java:
```diff
@@ -0,0 +1,26 @@
+        return 0;
+    }
+
+    SortedTopDocs rerank(List<SortedTopDocs> sortedTopDocs);
```
This API has changed. It still returns a `SortedTopDocs` but the docs are an extension used for ranking that contain additional information including positions. For full explanation we can still return a `SortedTopDocs`, but not cut down to `size` (I think).

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -59,6 +63,65 @@ public class QueryPhase {
+                    new DocValueFormat[0]
+                );
+        }
+
```
Also completely changed this. The `RankShardContext` is now responsible for controlling the flow of searches executed during the query phase. The only change to query phase is there is a path for standard execution and a path to defer to the `RankShardContext` instead.

jdconrad commented:
@benwtrent Thank you for the original review. 

I have tried to update this with a lot more structure based on your feedback with a set of base classes and a set of rrf classes implementing those base classes.

The base classes:

* `RankBuilder` - outer wrapper for parsing rank dsl
* `RankContextBuilder` - base class to extend to parse specific ranking such as `rrf` :)
* `RankContext` - base class to extend to rank on the coordinator
* `RankResult` - base class that extends `ScoreDoc` to extend that contains results from ranking on the coordinator
* `RankShardContext` - base class to extend to execute the required searches during the query phase on the shard
* `RankShardResult` - base class to extend that contains results from the query phase on the shard

WIP:
1. tests
2. documentation
3. registry?
4. should rrf be a plugin?

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix wrapped weight

jdconrad committed:
update version protections to 8.8 for read/write

jdconrad committed:
update rank test version compatibility

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java:
```diff
@@ -0,0 +1,52 @@
+            QueryPhase.executeInternal(rrfRankSearchContext);
+
+            List<TopDocs> rrfRankResults = new ArrayList<>();
+            rrfRankSearchContext.windowSize(windowSize);
```
What if `windowSize` is smaller than size?

jdconrad replied:
Let me double check on the behavior. I assume it will only return the amount of results it can find.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java:
```diff
@@ -0,0 +1,52 @@
+    public void executeQueries(SearchContext searchContext) {
+        try {
+            RRFRankSearchContext rrfRankSearchContext = new RRFRankSearchContext(searchContext);
+            QueryPhase.executeInternal(rrfRankSearchContext);
```
I may be missing something. But this seems way too expensive. We run the queries a minimum of 3 times. This initial execute should only be for aggs (no rescore phase) and it should be a filter query.

jdconrad replied:
You're not missing anything! I was just still considering the conditions under which this query should run. Aggregations only makes sense as we should probably restrict rank to no suggestions for now. (I think that's really complicated.) I thought by setting size to 0 it naturally became a filter query because the topdocscollector doesn't require score. If that's not the case I guess this query should be wrapped.

jdconrad replied:
I updated this to only run when we have suggestions, aggregations, or hit tracking.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java:
```diff
@@ -0,0 +1,52 @@
+            for (Query query : queries) {
+                rrfRankSearchContext.rrfRankQuery(query);
+                QueryPhase.executeInternal(rrfRankSearchContext);
+                rrfRankResults.add(rrfRankSearchContext.queryResult().topDocs().topDocs);
```
We should do a merge here with rank position in the sets. Right now this returns `windowSize * queries.size()` number of docs. It really only needs to be topDocs `windowSize` with their rank metadata

jdconrad replied:
I can do that. I assumed for correctness it was better to send all the results back and only do a single ranking. If this is the case we can probably just only have a single top docs and may not need changes to `QueryShardResult` at all.

jdconrad replied:
I split this up and redid the ranking. `window_size` now applies to both the shard and the coordinator.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankResult.java:
```diff
@@ -0,0 +1,59 @@
+    public final int[] positions;
+    public final float[] scores;
+
+    public RRFRankResult(int doc, int shardIndex, int size) {
```
Would be neat if we knew the the position group. 

Also, what about when a doc is in one set, but not another?

jdconrad replied:
I need to add some documentation here for sure. `positions` is always the size of the number of queries executed (excluding the aggs query). The positions are 1...n, so any document that isn't part of a query's result set is just naturally 0. Maybe -1 would be better to be more explicit?

jdconrad replied:
This is the same for now.

benwtrent commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -77,6 +78,11 @@ public final class SearchHit implements Writeable, ToXContentObject, Iterable<Do
+    private static final int NO_RANK = -1;
+    private int rank = NO_RANK;
+
+    private RankResult rankResult;
```
Rank should be part of `RankResult`

jdconrad replied:
I was wondering if we should only use `RankResult` for explain or not and just have a simple _rank otherwise? If not, I will remove rank and only have `RankResult`.

jdconrad replied:
I removed this completely for now. Based on the precedent already set, it seems like this information is better conveyed as part of an explanation in a follow up. Now only a single int for rank exists.

benwtrent commented on server/src/main/java/org/elasticsearch/search/SearchService.java:
```diff
@@ -1168,6 +1168,13 @@ private void parseSource(DefaultSearchContext context, SearchSourceBuilder sourc
         SearchExecutionContext searchExecutionContext = context.getSearchExecutionContext();
         context.from(source.from());
         context.size(source.size());
+        if (source.rank() != null) {
```
we need to validate that there aren't any `sorts` provided and throw if there are.

jdconrad replied:
Allowing sort actually doesn't break anything. But the behavior is certainly odd as it will apply to all queries. So I guess this should be part of the restrictions list.

jdconrad replied:
I restricted a number of things including sort. Since we do part of the ranking on the shard, we no longer pass FieldDoc to the coordinator, so we need to re-think about how to appropriately support sort. I think we may actually be able to replace RankDoc with FieldDoc in the future, but I think that is best saved for a follow up.

benwtrent reviewed (commented):
I think this overall approach better. I think there is some logic in the QueryPhase that needs reworking (running the query 3 times always seems no bueno).

Also, we need to handle:

 - windowSize being smaller than `size`
 - Handling when the doc is in one set but not another
 - Doing ranking on the shard, we are serializing way too many docs back. We only need the top set with some metadata about their set positions & scores

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -578,7 +593,12 @@ static ReducedQueryPhase reducedQueryPhase(
-        final SortedTopDocs sortedTopDocs = sortDocs(isScrollRequest, bufferedTopDocs, from, size, reducedCompletionSuggestions);
+        final SortedTopDocs sortedTopDocs = rankContext == null
+            ? sortDocs(isScrollRequest, bufferedTopDocs, from, size, reducedCompletionSuggestions)
+            : rankContext.rank(queryResults.stream().map(SearchPhaseResult::queryResult).toList(), topDocsStats);
```
The `ScoreDoc`s as part of `SortedTopDocs` in this case should be instances of `RankResult`. Given the `RankContext` is the only thing that is currently able to process rank information, I thought it made the most sense to do a cast with the `decorateSearchHit` method and possibly add other methods to the `RankContext` if we need this information elsewhere. There is already a precedent for this type of design when handling sort because we check for instanceof `FieldDoc` and cast as well.

I'm definitely open to changing this, but I don't have a good feel for either what should become part of `SortedTopDocs` and where else it would be used.

jdconrad replied:
I think for the first iteration this should continue to stay as it is because it requires no modifications to the fetch phase at this time, and we are already modifying a lot of the search phases code.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,121 @@
+import java.io.IOException;
+import java.util.Objects;
+
+public class RankBuilder implements Writeable, ToXContent {
```
Agreed. Switching to `NamedWriteable` will remove the need for this class.

jdconrad replied:
Switched this. It's definitely better.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,95 @@
+public abstract class RankContextBuilder implements Writeable, ToXContent {
+
+    protected final List<QueryBuilder> queryBuilders;
+    protected int size = 10;
+    protected int from = 0;
+
```
I will change this.

jdconrad replied:
Changed.

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -257,6 +261,9 @@ public SearchSourceBuilder(StreamInput in) throws IOException {
             }
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0)) {
+            rankBuilder = in.readOptionalWriteable(RankBuilder::new);
```
I will change this.

jdconrad replied:
Changed.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContextBuilder.java:
```diff
@@ -0,0 +1,159 @@
+
+    @Override
+    public String toString() {
+        return "RRFRankContextBuilder{" + "windowSize=" + windowSize + ", rankConstant=" + rankConstant + '}';
```
Missed as part of auto-generate. I will redo this.

jdconrad committed:
convert RankContextBuilder to a NamedWriteable

jdconrad committed:
fix parsing for namedwriteable

jdconrad committed:
spotless

jdconrad committed:
update main query to only execute for aggs, suggest, or total hits

jdconrad committed:
rename rank result to rank doc

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
partially moved ranking to query phase

jdconrad committed:
complete split of scoring across shard and coordinator

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
add restrictions

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/QueryPhaseResultConsumer.java:
```diff
@@ -90,7 +92,8 @@ public QueryPhaseResultConsumer(
 
         SearchSourceBuilder source = request.source();
-        this.hasTopDocs = source == null || source.size() != 0;
+        this.rankContext = source == null || source.rank() == null ? null : source.rank().rankContextBuilder().build();
```
Added a restriction for this.

jdconrad commented:
@benwtrent I think I have addressed all of the feedback from the past review. Would you please take another pass when you have time :) ? I'm going to start working on documentation and testing.

jdconrad committed:
fix optional named writeable

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-3

jdconrad commented:
@elasticmachine run elasticsearch-ci/part-3

jdconrad committed:
Add JavaDoc for base rank classes.

jdconrad committed:
add javadocs and comments for new rrf classes

jdconrad committed:
spotless

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContext.java:
```diff
@@ -0,0 +1,162 @@
+        // score if we already saw it as part of a previous query's
+        // doc set, otherwise we make a new doc and calculate the
+        // initial score
+        Map<String, RRFRankDoc> results = new HashMap<>();
```
Use a composite record class or something instead of concatenating strings.

https://richardstartin.github.io/posts/5-java-mundane-performance-tricks#use-wrappers-for-composite-hashmap-keys

Also, it would be good pre-size this hashmap if possible.

jdconrad replied:
Would you pre-size it to the max possible value every time?

jdconrad replied:
Implemented a record here for the key.

benwtrent replied:
> Would you pre-size it to the max possible value every time?

I would size it at least to something like the number of docs in the largest set of results. I don't think we want to dramatically OVERSIZE as a complete disjunction between the sets seems weird.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContextBuilder.java:
```diff
@@ -0,0 +1,207 @@
+    public ActionRequestValidationException validate(ActionRequestValidationException validationException, SearchSourceBuilder source) {
+        if (source.size() >= windowSize) {
+            validationException = addValidationError(
+                "[window size] must be greater than or equal to [size] for [rrf]",
```
```suggestion
                "[window_size] must be greater than or equal to [size] for [rrf]",
```

jdconrad replied:
Updated.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContextBuilder.java:
```diff
@@ -0,0 +1,207 @@
+            );
+        }
+        if (rankConstant < 1) {
+            validationException = addValidationError("[rank constant] must be greater than [0] for [rrf]", validationException);
```
```suggestion
            validationException = addValidationError("[rank_constant] must be greater than [0] for [rrf]", validationException);
```

jdconrad replied:
Updated.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContextBuilder.java:
```diff
@@ -0,0 +1,207 @@
+
+    public static final String NAME = "rrf";
+
+    public static final int RANK_CONSTANT_DEFAULT = 20;
```
Why 20? I might be missing something...

jdconrad replied:
It's funny I was chatting with @giladgal just this morning about this. Should be 60, correct? I will fix this.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankDoc.java:
```diff
@@ -0,0 +1,89 @@
+ */
+public class RRFRankDoc extends RankDoc {
+
+    public static final int NO_RANK = 0;
```
Seems like `-1` would be better? Keeping things `0'd` indexed for consistency is nice.

jdconrad replied:
Done.

benwtrent commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java:
```diff
@@ -0,0 +1,158 @@
+        // if a doc isn't part of a result set its position will be NO_RANK [0] and
+        // its score is [0f]
+        int queries = rrfRankResults.size();
+        Map<Integer, RRFRankDoc> docsToRankResults = new HashMap<>();
```
I think we could set expected size for this hashmap.

jdconrad replied:
Would you just have it be queries*window_size?

benwtrent replied:
`queries*window_size` I think so. `Maps.hashMapWithExpectedSize` or whatever its called

benwtrent commented on server/src/test/java/org/elasticsearch/action/search/RankSearchSingleNodeTests.java:
```diff
@@ -0,0 +1,104 @@
+import java.io.IOException;
+import java.util.List;
+
+public class RankSearchSingleNodeTests extends ESSingleNodeTestCase {
```
We should also have some multi-node tests. I bet those will find some very interesting problems

jdconrad replied:
Added several multi-node tests that include variations of different combinations of bm25 and knn queries along with aggregations.

benwtrent reviewed (commented):
I like the direction of things. 

We do need some more tests:

 - serialization & parsing tests
 - multi-node tests
 - etc.

But it is looking good!

jdconrad committed:
stub out documentation for rrf

jdconrad commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -606,6 +623,8 @@ public static class Fields {
         static final String _PRIMARY_TERM = "_primary_term";
         static final String _SCORE = "_score";
+        static final String _RANK = "_rank";
+        static final String _RANK_RESULT = "_rank_result";
```
Fixed. This was for adding the additional position information, but that should become part of explain instead.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,128 @@
+    public int hashCode() {
+        return Objects.hash(queryBuilders, size, from);
+    }
+
```
Thanks for catching this. I messed this up when I removed the extraneous outer builder in favor of registering `RankContextBuilder` as a named x-content object.

jdconrad replied:
I have cleaned this up.

jdconrad committed:
clean up writeTo and toXContent for RankContextBuilder

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,128 @@
+        from = in.readVInt();
+    }
+
+    public void writeTo(StreamOutput out) throws IOException {
```
Done.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,128 @@
+        if (o == null || getClass() != o.getClass()) return false;
+        RankContextBuilder that = (RankContextBuilder) o;
+        return size == that.size && from == that.from && Objects.equals(queryBuilders, that.queryBuilders);
+    }
```
Done.

jdconrad committed:
make NO_RANK -1

jdconrad committed:
update search hit tests

jdconrad committed:
remove dead code

jdconrad committed:
separate internal information for ranking into RankContextInternal and
ShardSearchRequest

jdconrad committed:
add wire tests for RankContextInternal

jdconrad committed:
serialization tests for RankContextBuilder

jdconrad committed:
add serialization tests for RRFRankDoc

jdconrad committed:
more serialization tests

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
start to add integ tests for rrf

jdconrad committed:
add framework for rrf rank integration tests and a couple of basic tests

jdconrad committed:
remove single node tests as they are replace by integration tests

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
add descending vectors for tests

jdconrad committed:
add additional tests including multi knn and bm25 and tests with aggs

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix comment

jdconrad committed:
spotless

jdconrad committed:
fix consistency in rrf it tests

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
add it tests for rrf involving scripted metric aggs

jdconrad committed:
add serialziation tests for QuerySearchResult

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
add validation tests to SearchRequest

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
update random search request generator to include rank

not-napoleon reviewed (commented):
I only looked at the surface where this intersects my work in SearchContext and related classes, but from that perspective this all looks fine.  Thanks for tagging me!

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
add hash map pre sizing

jdconrad committed:
Merge branch 'main' into rrf9

javanna commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankDoc.java:
```diff
@@ -0,0 +1,116 @@
+ * {@code RRFRankDoc} supports additional ranking information
+ * required for RRF.
+ */
+public class RRFRankDoc extends RankDoc {
```
Similar to the comment I left on RankDoc, I am hoping that we can do without this class

jdconrad replied:
My response is similar to the one for RankDoc as well :) . I do think this could be replaced by FieldDoc, but given the complexity would be best served in a follow up PR.

javanna reviewed (commented):
I did a first pass and left an initial batch of comments, I will continue tomorrow focusing on the multiple queries execution, the filtered search context and the re-ranking on the coord node. All my comments are mostly cosmetic, no blockers, it all looks quite clean and thought through to me!

jdconrad committed:
Update docs/changelog/93396.yaml

elasticsearchmachine commented:
Hi @jdconrad, I've created a changelog YAML for you. *Note* that since this PR is labelled `release highlight`, you need to update the changelog YAML to fill out the extended information sections.

jdconrad commented:
@javanna Thank you for the feedback! It's going to take me some time to go through these as we discussed already a bit.

jdconrad committed:
update changelog description

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix changelog

jdconrad committed:
fix tests

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
basic xpack plugin outline

jdconrad committed:
start of plugin

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix configuration issues with plugin

jdconrad committed:
move more unit tests

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
Merge branch 'main' into rrfplugin

jdconrad committed:
remove rank context internal wrapper

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextInternal.java:
```diff
@@ -0,0 +1,87 @@
+ * Holds internal information about the queries
+ * used for ranking required for shard phases.
+ */
+public class RankContextInternal implements Writeable, Rewriteable<RankContextInternal> {
```
I agree. I initially thought it would be more complex so that additional abstraction would be useful. I have removed the abstraction in favor of your suggestion here. The only thing I didn't do was add an additional method to SearchExecutionContext because it returns ParsedQuery for its toQuery methods, so I thought that might be confusing. I just added a simple loop where we do the conversion instead.

jdconrad committed:
rename RankContextBuilder to RankBuilder

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,145 @@
+ * of ranking input, but instead generated from their respective values within a search
+ * query. We cannot generate the list of queries until after any kNN queries are executed.
+ */
+public abstract class RankContextBuilder<RCB extends RankContextBuilder<RCB>> implements VersionedNamedWriteable, ToXContentObject {
```
I changed this.

jdconrad committed:
rename RankContext to RankCoordinatorContext

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Generates a context used to perform global ranking on the coordinator.
+     */
+    public abstract RankContext build(int size, int from);
```
Done!

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,145 @@
+    }
+
+    public String toString(Params params) {
+        try {
```
I was using behavior similar to other builders. The nice thing about this is the subclasses get it for "free" since they'll also need they're own x-content methods.

javanna replied:
doesn't String.toString do the same including handling the exception? My thinking was to further simplify the method, but maybe I am missing why that is not possible

jdconrad replied:
Oh, I assumed it was different. Let me double check this.

jdconrad committed:
move multiple query execution for ranking into QueryPhase

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,145 @@
+    /**
+     * Generates the query used for aggregations and suggesters.
+     */
+    public abstract QueryBuilder searchQuery(List<QueryBuilder> queryBuilders);
```
Done!

jdconrad committed:
move rank query execution into query phase

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankContextBuilder.java:
```diff
@@ -0,0 +1,145 @@
+
+    public static final int DEFAULT_WINDOW_SIZE = SearchService.DEFAULT_SIZE;
+
+    protected int windowSize = DEFAULT_WINDOW_SIZE;
```
This should probably be changed to standard ObjectParser instead of Constructing. Seems like most of the builders allow configuration outside of construction which is used during testing.

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -59,6 +59,14 @@ public class QueryPhase {
+        }
+    }
+
+    public static void executeInternal(SearchContext searchContext) throws QueryPhaseExecutionException {
```
I renamed this to executeQuery. It's still a bit more complicated than previously, but all the logic is now contained in QueryPhase instead of externally which seems like a big improvement.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankShardContext.java:
```diff
@@ -0,0 +1,158 @@
+        }
+    }
+
+    protected void sort(List<TopDocs> rrfRankResults, QuerySearchResult querySearchResult) {
```
I think there may be a path forward here as well, but I would like to change this when working directly in a branch with a plugin to see how it may best fit.

jdconrad committed:
rename method

javanna commented on server/src/main/java/org/elasticsearch/search/rank/rrf/RRFRankContextBuilder.java:
```diff
@@ -0,0 +1,153 @@
+
+    private static final ConstructingObjectParser<RRFRankContextBuilder, Void> PARSER = new ConstructingObjectParser<>("rrf", args -> {
+        RRFRankContextBuilder builder = new RRFRankContextBuilder();
+        builder.windowSize(args[0] == null ? DEFAULT_WINDOW_SIZE : (int) args[0]);
```
I think it looks weird and intuitively, users would expect it all under RRF. Moving it at the top-level feels like would be to reflect the classes hierarchy which users are not interested in.

jdconrad replied:
I will leave it as is.

jdconrad committed:
fix test

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix npe

jdconrad committed:
Merge branch 'rrf9' into rrfplugin

jdconrad committed:
move most of core rrf classes to plugin

jdconrad committed:
spotless

jdconrad committed:
fix test dependencies

jdconrad committed:
add yaml tests

jdconrad commented:
@javanna I have started a branch here moving rrf into a plugin. I have a good chunk of the tests working there, but I'm still thinking about tests for validation in server and how to reduce the scope of the API extensions points in the rank contexts.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankShardContext.java:
```diff
@@ -0,0 +1,40 @@
+     * global rank method must be managed and executed when this
+     * is called.
+     */
+    public abstract void executeQueries(SearchContext searchContext);
```
I moved this code to QueryPhase with the assumption that all future rankers will use the same pattern of querying for ranking.

jdconrad committed:
add test rank objects for testing validation in server

jdconrad committed:
spotless

jdconrad committed:
fix test

jdconrad committed:
change signature for rrf rank shard context

jdconrad committed:
fix yaml test directory structure

jdconrad committed:
add platinum license and fix yaml tests

jdconrad committed:
avoid license check in parsing tests for builder

jdconrad committed:
switch constructing object parser to standard object parser

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
Merge branch 'rrf9' into rrfplugin

jdconrad commented:
@javanna I have merged the plugin branch into this PR. This is ready for another round of review. Thanks!

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix source builder test

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
fix test

jdconrad committed:
add invalid license test to yaml

jdconrad committed:
start to move scripted metric agg tests to yml

jdconrad committed:
complete moving scripted metric tests to yml

jdconrad committed:
Merge branch 'main' into rrf9

javanna commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -59,6 +64,71 @@ public class QueryPhase {
+                    break;
+                }
+                rankSearchContext.rankQuery(query);
+                QueryPhase.executeQuery(rankSearchContext);
```
it seems to me that we don't support any feature within executeQuery other than what executeWithCollectors does, hence we could call such method directly for clarity?

jdconrad replied:
I think that's true for now. Obviously, this may change as we add more feature support for RRF, but I'll go ahead and make this change.

javanna replied:
Agreed, we will have to change it but then it's going to be more visible that we need to do so.

javanna commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -92,7 +163,7 @@ public static void execute(SearchContext searchContext) throws QueryPhaseExecuti
      * wire everything (mapperService, etc.)
      */
-    static void executeInternal(SearchContext searchContext) throws QueryPhaseExecutionException {
+    static void executeWithCollectors(SearchContext searchContext) throws QueryPhaseExecutionException {
```
I am not sold on the new name here, because the collectors get created later on as part of the searchWithCollector method. I must admit I cannot come up with anything better than executeQueryInternal so far :)

jdconrad replied:
I think this is similar to what it was called before? But I admit the name isn't great here. Open to suggestions :)

jdconrad replied:
I changed this to `addCollectorsAndExecuteQuery`

javanna commented on server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java:
```diff
@@ -0,0 +1,97 @@
+ * a single time with new results. The results must be collected out of
+ * {@link QuerySearchResult} after each execution.
+ */
+public class RankSearchContext extends FilteredSearchContext {
```
I was wondering if we could use the original search context directly for the initial aggs search and this one only for the different rank searches. I feel like that would remove the null checks below and simplify things a bit.

For instance,  the aggregations method could simply throw, as well as all the features that the rank searches don't support (yet).

jdconrad replied:
I went ahead and just outright removed the unused methods altogether. I think that's simpler than trying to throw for the unused methods.

javanna commented on server/src/main/java/org/elasticsearch/search/rank/RankShardContext.java:
```diff
@@ -0,0 +1,52 @@
+        return size;
+    }
+
+    public int from() {
```
are getters for these needed if they are never used?

jdconrad replied:
Removed.

javanna commented on server/src/main/java/org/elasticsearch/search/rank/RankShardContext.java:
```diff
@@ -0,0 +1,52 @@
+        return windowSize;
+    }
+
+    public abstract RankShardResult sort(List<TopDocs> rrfRankResults);
```
can we add javadocs? That will also answer my question on "why is this needed again?" . This is to reduce the number of docs that we return back to the coordinating node, right?

jdconrad replied:
Added :)

javanna commented on docs/changelog/93396.yaml:
```diff
@@ -0,0 +1,51 @@
+    ----
+
+    The above example will execute two separate queries for the match query
+    and the kNN query. It will preserve separate result sets up the point
```
From this explanation, as an advanced user I may wonder where the separate queries are executed, how this affects the search roundtrips. Instead of expanding this though, I wonder if we should keep the highlight description more high-level and rather explain what RRF helps with: poorly said,  it mixes together docs that have scores on different scales?

jdconrad replied:
Docs need work for sure. I've still only stubbed out the RRF docs.

javanna commented on docs/changelog/93396.yaml:
```diff
@@ -0,0 +1,51 @@
+    results sets together with `sum(1/(k+d))` where k is a ranking constant
+    and d is a document's scored position within a result set from a query.
+
+    The API for this change adds a `rerank` top-level element to the search
```
the section is called rank and not rerank?

jdconrad replied:
Same comment as above.

jdconrad replied:
Fixed.

javanna commented on docs/reference/search/search.asciidoc:
```diff
@@ -550,6 +550,18 @@ Period of time used to extend the life of the PIT.
 
+[[request-body-rank]]
+`rank`::
+(Optional, object) Defines a method for combining and ranking multiple queries
```
I wonder if mentioning queries may be confusing, because we currently only allow to combine one or more knn with one query. I am merely referring to the fact that users may not see knn as a query because it's a different section of the search API.

jdconrad replied:
Agreed. I will change when I work on the docs in the last step.

jdconrad replied:
Updated.

javanna commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -338,6 +345,9 @@ public void writeTo(StreamOutput out) throws IOException {
         }
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0)) {
+            out.writeOptionalNamedWriteable(rankBuilder);
+        }
```
I would do this for safety in all places where we are serializing additional info. in some internal places it could even be an assert though probably.

rjernst replied:
I do not think we should have validation exceptions like that in the middle of serializing objects. This is because I think we should move towards less special casing inside StreamInput/Output code, so that we could eventually generate the code base on simple POJOs (or better yet, records). The writeTo and StreamInput constructors are meant as simple serialization methods, not validators.

There are two scenarios you described: mixed cluster and older ccs.

For mixed cluster, I do not think we should do any special checking. That scenario means a user is sending a request asking for rrf *in the middle of upgrading to 8.8*. It's an edge case which we protect users from by suggesting first upgrading clusters before clients.

For CCS, the scenario is valid. However, we should be able to check such scenarios elsewhere, sending back an appropriate 400, not an internal error because we failed to write a stream. I'm not sure the best place to do that, but perhaps it would be in `TransportSearchAction.executeRequest`, similar to the `checkCCSVersionCompatibility` call there?

javanna replied:
We have switched to a different pattern of throwing exception a while ago, and I thought it was a discussion we had with the overall ES team. This came out of the CCS working group I believe, and it affects both mixed clusters and CCS, where CCS is like you identified the most prominent case.

In general, we don't want to accept requests that are asking for things that we can't accommodate. We did it for runtime fields: if you provide for runtime mappings in a search request, we will fail serializing the request to an older version node that does not support it, instead of doing what the user asked only on nodes that support it and silently failing on older nodes. The checkCCSVersionCompatibility builds on it as it relies on the serialization layer throwing exception whenever something that the previous version does not support gets serialized.

@DaveCTurner @nik9000 I remember you two were involved in these discussions too, and I initially had a similar reaction as Ryan, but then I got convinced and I adopted the "new approach". Do you have more context to it?

DaveCTurner replied:
I definitely prefer to detect the use of "future" features, and reject them, before we even get to the point of trying to serialize a message that the remote node won't understand. We can give much more useful feedback (and avoid unnecessary work) if we do.

But we still have to do _something_ on the serialization path when we encounter something that can't be translated into an older protocol version. Either we bail out with an exception, or we just proceed blindly and send an invalid message. Given this choice, I think we should prefer to bail out wherever possible. Receiving an invalid message trips assertions and throws an exception (on the sender) but there are cases where the message remains syntactically valid, yet differs semantically from what we intended, and that's a bug.

If we have upstream checks that prevent it then we can certainly assert that this branch never happens too, but still IMO we should throw an exception after the `assert false` just in case we're missing test coverage.

rjernst replied:
I understand the concern, and agree it is better to bail. Adding assertions/throw assertions in the else case sounds good. My main ask is to keep validation separate from serialization. If we can move such a check earlier in the request, such as in `TransportSearchAction.executeRequest`, I think it will make the code easier to reason about too.

jdconrad replied:
@javanna I went ahead and added `IllegalArgumentException`s at the different places where we check for 8.8. Would you please point me to where runtime fields does this check? I see the one for CCS, but I don't see one for just a standard mixed cluster.

javanna replied:
This is the check that I meant: https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java#L321

javanna reviewed (commented):
I did another round of review. I am only 70/80% done but I have enough comments to submit this round and continue it tomorrow. This is looking great!

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+ * of ranking input, but instead generated from their respective values within a search
+ * query. We cannot generate the list of queries until after any kNN queries are executed.
+ */
+public abstract class RankBuilder<RCB extends RankBuilder<RCB>> implements VersionedNamedWriteable, ToXContentObject {
```
I found this to be common pattern in our builders in general. Seems nice for avoiding casting, but I'm happy to remove the generic type if you want.

javanna replied:
True, it is one of those common patterns that I find misuse generics :) How does it help in testing? I may have missed that.

jdconrad replied:
For integration tests, it's nice to set window_size and rank_constant, but I think you're right that in a comment below -- this can all effectively be part of an immutable object through a constructor. I will give that a try.

jdconrad replied:
Removed the generic type.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+    }
+
+    public String toString(Params params) {
+        try {
```
Oops! I meant to change this one. Thank you for catching it for the second time. :)

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+ * and subsequent generation of appropriate contexts for handling searches that
+ * require multiple queries for global rank relevance.
+ *
+ * This class contains a {@code List<QueryBuilder>}s, size, and from members that
```
Good catch! I removed the second paragraph which is no longer relevant since the actual queries have been moved elsewhere.

jdconrad commented on x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilder.java:
```diff
@@ -0,0 +1,126 @@
+            validationException = addValidationError("[rank_constant] must be greater than [0] for [rrf]", validationException);
+        }
+        if (source.knnSearch().isEmpty() || source.query() == null && source.knnSearch().size() < 2) {
+            validationException = addValidationError("[rrf] requires a minimum of [2] queries", validationException);
```
I added additional information to this error message. Hopefully it's a bit clearer now as to what's required.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+
+    public static final int DEFAULT_WINDOW_SIZE = SearchService.DEFAULT_SIZE;
+
+    protected int windowSize = DEFAULT_WINDOW_SIZE;
```
I made this private.

I switched to an ObjectParser because it's a better pattern for a Builder whereas ConstructingObjectParser is a better fit for an immutable object.

jdconrad replied:
I think it can be immutable if we move the values to the constructor. I'll figure out if there's any reason to not do this.

jdconrad replied:
`RankBuilder` and `RRFRankBuilder` now immutable.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+        doWriteTo(out);
+    }
+
+    public abstract void doWriteTo(StreamOutput out) throws IOException;
```
I have the rrf classes in a different package, but maybe it makes more sense to move them all into `rank` instead of `rank.rrf`?

javanna replied:
protected is also good.

jdconrad replied:
Ah, of course. For some reason I read this as private or package-protected which doesn't make sense.

jdconrad replied:
Done.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+    public abstract ActionRequestValidationException validate(
+        ActionRequestValidationException validationException,
+        SearchSourceBuilder source
+    );
```
I moved this exception to RankBuilder for the minimum query count, but I think it still fits better to have the rank constant be a validation exception since it should be caught as part of the original request. To accommodate this, I've added a doValidate in a similar fashion to doEquals and doHashCode.

javanna replied:
Thanks. Validate is convenient, but ideally you can fail fast as soon as you set something that is wrong, and you see its validation code directly in the setter. Validate is still useful for places where you need to check more than one value and you need the object to be fully initialized.

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -390,6 +400,15 @@ public List<KnnSearchBuilder> knnSearch() {
         return Collections.unmodifiableList(knnSearch);
     }
 
+    public SearchSourceBuilder rankContextBuilder(RankBuilder<?> rankBuilder) {
```
Yes! Good catch. Fixed.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchRequestBuilder.java:
```diff
@@ -190,6 +191,14 @@ public SearchRequestBuilder setMinScore(float minScore) {
+    /**
+     * Defines a rank method for ranking results.
+     */
+    public SearchRequestBuilder setRankContextBuilder(RankBuilder<?> rankBuilder) {
```
Also fixed.

jdconrad commented on test/framework/src/main/java/org/elasticsearch/search/rank/RankBuilderTests.java:
```diff
@@ -0,0 +1,37 @@
+
+import java.io.IOException;
+
+public abstract class RankBuilderTests<RCB extends RankBuilder<RCB>> extends AbstractXContentSerializingTestCase<RCB> {
```
I went ahead and remove the base class here. I agree it doesn't add any value.

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1259,6 +1279,21 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
+                            "unexpected start token [" + token + "] for [" + RANK_FIELD.getPreferredName() + "]"
+                        );
+                    }
+                    rankBuilder = parser.namedObject(RankBuilder.class, parser.currentName(), null);
```
I added a line that uses the name of the RankBuilder so in this case `rrf`. I agree this makes more sense than just straight `rank` which may not be as helpful in the future.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -361,6 +361,50 @@ public ActionRequestValidationException validate() {
+                if (scroll) {
+                    validationException = addValidationError("[rank] cannot be used in a scroll context", validationException);
+                }
+                if (source.size() < -1 || source.size() == 0) {
```
Ah I see! I'll remove this check. Thanks.

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -361,6 +361,50 @@ public ActionRequestValidationException validate() {
+                if (source.size() < -1 || source.size() == 0) {
+                    validationException = addValidationError("[rank] requires [size] greater than [0]", validationException);
+                }
+                if (source.size() > source.rankContextBuilder().windowSize()) {
```
Yes! That was super subtle. I have updated this to use default size if it's `-1`.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -225,6 +230,7 @@ public ShardSearchRequest(
         this.numberOfShards = numberOfShards;
         this.searchType = searchType;
         this.source(source);
+        this.rankQueryBuilders = rankQueryBuilders == null ? List.of() : rankQueryBuilders;
```
This is true. I went ahead and removed it from the constructor altogether and just set it to `List.of()` initially.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -414,6 +429,14 @@ public SearchSourceBuilder source() {
+    }
+
+    public void rankQueryBuilders(List<QueryBuilder> rankQueryBuilders) {
+        this.rankQueryBuilders = rankQueryBuilders == null ? List.of() : rankQueryBuilders;
```
Nope. Removed the null check.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankShardContext.java:
```diff
@@ -0,0 +1,52 @@
+public abstract class RankShardContext {
+
+    protected final List<Query> queries;
+    protected final int size;
```
It is not. I removed it.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Generates a context used to execute required searches on the shard.
+     */
+    public abstract RankShardContext build(List<Query> queries, int size, int from);
```
Changed.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankBuilder.java:
```diff
@@ -0,0 +1,140 @@
+    /**
+     * Generates a context used to perform global ranking on the coordinator.
+     */
+    public abstract RankCoordinatorContext build(int size, int from);
```
Changed.

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -59,6 +64,71 @@ public class QueryPhase {
+            if (searchContext.suggest() != null
+                || searchContext.trackTotalHitsUpTo() != TRACK_TOTAL_HITS_DISABLED
+                || searchContext.aggregations() != null) {
+                QueryPhase.executeQuery(rankSearchContext);
```
It needs the `rankSearchContext` as we intentionally set `size` to `0` to avoid scoring on aggregations in this case.

jdconrad committed:
response to pr comments

jdconrad commented:
@javanna Thank you for the latest round of review! I have updated the PR with the majority of your suggestions. Two notes: I asked @rjernst to comment on the exception in writeable because I remember him having a recent conversation around this topic, and I know I need to work on the docs, so I haven't updated anything around your feedback there, yet.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -344,6 +356,9 @@ protected final void innerWriteTo(StreamOutput out, boolean asKey) throws IOExce
         out.writeOptionalWriteable(source);
+        if (out.getTransportVersion().onOrAfter(TransportVersion.V_8_8_0)) {
+            out.writeNamedWriteableList(rankQueryBuilders);
+        }
```
Please see the comment below from @rjernst

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-1

javanna commented on x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorContext.java:
```diff
@@ -0,0 +1,164 @@
+    public void decorateSearchHit(ScoreDoc scoreDoc, SearchHit searchHit) {
+        assert scoreDoc instanceof RRFRankDoc;
+        RRFRankDoc rankResult = (RRFRankDoc) scoreDoc;
+        searchHit.setRank(rankResult.rank);
```
not too sure about positions (I asked above what it holds), but rank could be seen as something general enough for every ranker? Double checking that we need this method, maybe we could instead get (from RankDoc) and set the rank directly where this method is currently called?

jdconrad replied:
Moved rank to `RankDoc` and removed `decorateSearchHit`.

javanna commented on server/src/main/java/org/elasticsearch/search/rank/RankDoc.java:
```diff
@@ -0,0 +1,65 @@
+ * Subclasses should extend this with additional information
+ * required for their global ranking method.
+ */
+public abstract class RankDoc extends ScoreDoc implements VersionedNamedWriteable {
```
I think that this could implement Writeable, does not need to go further?

jdconrad replied:
Fixed.

javanna commented on x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RankRRFPlugin.java:
```diff
@@ -0,0 +1,45 @@
+
+import java.util.List;
+
+public class RankRRFPlugin extends Plugin {
```
for consistency, should we call it RRFRankPlugin? Seems like all other classes in this package follow that convention.

jdconrad replied:
Done.

javanna commented on x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RankRRFPlugin.java:
```diff
@@ -0,0 +1,45 @@
+        return List.of(
+            new NamedWriteableRegistry.Entry(RankBuilder.class, NAME, RRFRankBuilder::new),
+            new NamedWriteableRegistry.Entry(RankShardResult.class, NAME, RRFRankShardResult::new),
+            new NamedWriteableRegistry.Entry(RankDoc.class, NAME, RRFRankDoc::new)
```
I believe that RankDoc does not need to be a named writeable as it's now serialized as part of RRFRankShardResult which is itself a named writeable

jdconrad replied:
Fixed.

javanna commented on server/src/test/java/org/elasticsearch/search/query/QueryPhaseTests.java:
```diff
@@ -951,7 +952,7 @@ public void testMinScore() throws Exception {
+        QueryPhase.addCollectorsAndSearch(context);
         assertEquals(10, context.queryResult().topDocs().topDocs.totalHits.value);
 
         reader.close();
```
Do you think it would be possible to add some unit tests to this class (QueryPhaseTests) around the QueryPhase changes you made? For instance, we could test the case where the initial search for aggs is or isn't performed depending on track_total_hits and aggs in the request. We could test the execution of the following queries that are kept separate etc.

javanna replied:
Do we need tests also for named queries and stats ( I figured that's what the node queue size and service time are used for)?

jdconrad replied:
Added a test to check if queries are executed in the expected order.

Any suggestions for stats?

javanna replied:
I'd try and verify that when the stats parameter is provided in combination with RRF the behaviour is the expected one. We could do this in a unit test around the actual execute method in QueryPhase, as well as an integration test that verifies the end-to-end behaviour.

javanna replied:
it's fine to leave this for a follow-up

javanna commented on server/src/test/java/org/elasticsearch/action/search/SearchPhaseControllerTests.java:
```diff
@@ -247,6 +247,7 @@ public void testMerge() {
                 0,
                 true,
                 InternalAggregationTestCase.emptyReduceContextBuilder(),
+                null,
```
would we want to unit test the new rank codepath here?

jdconrad replied:
Added a test for this.

javanna commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -136,40 +137,75 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+            // this path will keep knn queries separate for ranking per shard
+            // if there are multiple knn queries
+
+            List<QueryBuilder> rankQueryBuilders = new ArrayList<>();
```
should we expand testing in DFSQueryPhaseTests to cover for this additional codepath around keeping the different knn results separate?

jdconrad replied:
Done.

javanna commented on x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankCoordinatorContext.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    @Override
+    public SortedTopDocs rank(List<QuerySearchResult> querySearchResults, TopDocsStats topDocsStats) {
```
Could we have unit tests for this rank method?

jdconrad replied:
Added.

javanna commented on x-pack/plugin/rank-rrf/src/main/java/org/elasticsearch/xpack/rank/rrf/RRFRankShardContext.java:
```diff
@@ -0,0 +1,88 @@
+    }
+
+    @Override
+    public RRFRankShardResult sort(List<TopDocs> rrfRankResults) {
```
could we have unit tests for this sort method?

jdconrad replied:
Added.

javanna commented on x-pack/plugin/rank-rrf/src/test/java/org/elasticsearch/xpack/rank/rrf/RRFRankBuilderTests.java:
```diff
@@ -0,0 +1,61 @@
+        parser.nextToken();
+        Assert.assertNull(parser.currentToken());
+        return builder;
+    }
```
add two super simple test methods that verify the two methods that build contexts?

jdconrad replied:
Done.

javanna reviewed (commented):
I finished the review round that I started yesterday, still looking really good. Most comments are really around refining, questions, and adding tests. let's work on the docs and ship it?

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
Merge branch 'main' into rrf9

javanna commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -136,40 +137,75 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+            BoolQueryBuilder searchQuery = new BoolQueryBuilder();
+            for (QueryBuilder queryBuilder : rankQueryBuilders) {
+                searchQuery.should(queryBuilder);
+            }
```
good question we do that exclusively for aggs as they are run against the conjunction of all the matches.

jdconrad replied:
We use this as the primary query for executing aggregations or hit tracking if those features are required. We may be able to figure out when we don't need to do this, but since a `null` gets rewritten to a `match_all` I think there may be some more work required there.

jdconrad committed:
switch to constructingobjectparser

jdconrad committed:
remove generic from rank builder

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/SearchRequest.java:
```diff
@@ -361,24 +362,32 @@ public ActionRequestValidationException validate() {
-                }
-                if (source.size() < -1 || source.size() == 0) {
+            if (source.rankBuilder() != null) {
+                int size = source.size() == -1 ? SearchService.DEFAULT_SIZE : source.size();
```
I agree on cleaning this up. Let's do that after as you mentioned.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java:
```diff
@@ -0,0 +1,67 @@
+    @Override
+    public QuerySearchResult queryResult() {
+        return querySearchResult;
+    }
```
Named queries do seem to work because they're run during the fetch phase on the BM25 query right now.

javanna replied:
interesting. that makes sense though. I think throwing from all the methods that we don't expect to be called would be a good way to enforce we don't do the wrong thing.

jdconrad replied:
Done.

jdconrad commented on server/src/main/java/org/elasticsearch/search/internal/FilteredSearchContext.java:
```diff
@@ -178,6 +181,26 @@ public SearchContext fetchSourceContext(FetchSourceContext fetchSourceContext) {
+    @Override
+    public SearchContext fetchFieldsContext(FetchFieldsContext fetchFieldsContext) {
+        return in.fetchFieldsContext(fetchFieldsContext);
+    }
```
The other contexts that extend `FilteredSearchContext` all overrode them, so I think they got overlooked. Didn't really make sense to me to have these in the `RankSearchContext` so I added them here.

javanna replied:
++

jdconrad committed:
make modifier protected instead of public

jdconrad committed:
remove position from search hit

jdconrad commented on server/src/main/java/org/elasticsearch/search/SearchHit.java:
```diff
@@ -687,6 +717,14 @@ public XContentBuilder toInnerXContent(XContentBuilder builder, Params params) t
+            builder.field(Fields._RANK, rank);
+
+            if (positions.length > 0) {
+                builder.array(Fields._POSITIONS, positions);
```
`_score` will be `null` for ranking as the scores across documents can't be compared so adding/multiplying/etc. doesn't make sense to present.

`_rank` will be the rank from 1 to size for each ranked document. It won't be included as part of the search hit if ranking isn't used.

jdconrad committed:
add checks for serialization to an incompatible version

jdconrad committed:
add version checks in search request tests for rank

jdconrad committed:
clean up rank search context

jdconrad commented on server/src/main/java/org/elasticsearch/action/search/DfsQueryPhase.java:
```diff
@@ -136,40 +137,75 @@ private ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request)
+                    }
+                }
+                scoreDocs.sort(Comparator.comparingInt(scoreDoc -> scoreDoc.doc));
+                KnnScoreDocQueryBuilder knnQuery = new KnnScoreDocQueryBuilder(scoreDocs.toArray(new ScoreDoc[0]));
```
Agreed -- worth doing as a follow up.

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Override
+    public Query query() {
+        return rankQuery;
```
I went ahead and switched to extending `SearchContext` directly instead of `FilteredSearchContext`. I went through each method and either overrode specifically for ranking, called the original `DefaultSearchContext` method where is made sense, or throw `UnsupportedOperationException`. I agree that this should help significantly with future maintenance and additions.

javanna replied:
thanks!

jdconrad commented on server/src/main/java/org/elasticsearch/search/query/QueryPhase.java:
```diff
@@ -59,6 +64,69 @@ public class QueryPhase {
+            throw e;
+        } catch (Exception e) {
+            throw new QueryPhaseExecutionException(searchContext.shardTarget(), "Failed to execute main query", e);
+        }
```
We were re-writing the individual rank queries, but this has been moved to the standard query execution flow instead, so I removed this.

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
reset FilteredSearchContext

jdconrad committed:
add back rank shard context to filtered search context

jdconrad commented on server/src/main/java/org/elasticsearch/search/rank/RankShardContext.java:
```diff
@@ -0,0 +1,49 @@
+     * know any searches that match the same document must be on the same shard, we can sort on the shard
+     * instead for a top window_size set of results and reduce the amount of data we serialize.
+     */
+    public abstract RankShardResult sort(List<TopDocs> rrfRankResults);
```
Done.

jdconrad commented on server/src/main/java/org/elasticsearch/search/builder/SearchSourceBuilder.java:
```diff
@@ -1259,6 +1279,22 @@ private SearchSourceBuilder parseXContent(XContentParser parser, boolean checkTr
+                        );
+                    }
+                    parser.nextToken();
+                    searchUsage.trackSectionUsage(rankBuilder.getWriteableName());
```
Done.

jdconrad committed:
remove named writable from RankDoc

jdconrad committed:
make rank universal as part of rank doc

jdconrad committed:
rename rrf plugin

jdconrad committed:
test fix

jdconrad committed:
add query phase tests

jdconrad committed:
add search phase controller test for rank

jdconrad committed:
add dfs test for rank path

jdconrad committed:
add tests for creating the shard and coordinator context from the
builder

jdconrad committed:
add a unit test for combine method

jdconrad committed:
partial ranking test

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad committed:
add unit test for rrf rank coordinator context

jdconrad committed:
clean up change log

jdconrad committed:
more clean up of change log

javanna commented on server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java:
```diff
@@ -0,0 +1,564 @@
+    }
+
+    /**
+     * Ranking is now allowed with scroll.
```
s/now/not ?

javanna commented on server/src/main/java/org/elasticsearch/search/rank/RankSearchContext.java:
```diff
@@ -0,0 +1,564 @@
+
+    @Override
+    public SourceLoader newSourceLoader() {
+        throw new UnsupportedOperationException();
```
so many methods! :) would it make sense to move up all the methods that do something and down all the methods that do nothing or throw? Maybe add a big comment that says "from now on, nothing to look at"

jdconrad replied:
Fixed.

javanna reviewed (approved):
LGTM great work, thank you!

jdconrad committed:
clean up search doc rank api description

jdconrad committed:
response to pr comments.

jdconrad committed:
Merge branch 'main' into rrf9

jdconrad commented:
@javanna Thank you for the many, many rounds of review!

jdconrad commented:
@elasticsearchmachine run elasticsearch-ci/part-1
